### DISCUSSION TITLE
I am getting Error while calling Birt Report

### DISCUSSION THREAD

**Discussion by irfankundi786** [eclipse-birt/birt#1724] [Q&A]
**Error when creating service class using Birt Report to Run in Spring MVC. Here is Log of Error Description**
 `08-Jun-2024 21:06:11.763 INFO [http-nio-8080-exec-531] org.apache.catalina.core.ApplicationContext.log Destroying Spring FrameworkServlet 'dispatcher'
08-Jun-2024 21:06:11.764 INFO [http-nio-8080-exec-531] org.apache.catalina.core.ApplicationContext.log Closing Spring root WebApplicationContext
08-Jun-2024 21:06:27.899 INFO [http-nio-8080-exec-530] org.apache.catalina.core.ApplicationContext.log No Spring WebApplicationInitializer types detected on classpath
08-Jun-2024 21:06:27.931 INFO [http-nio-8080-exec-530] org.apache.catalina.core.ApplicationContext.log Initializing Spring root WebApplicationContext
08-Jun-2024 21:06:32.049 SEVERE [http-nio-8080-exec-530] org.apache.catalina.core.StandardContext.listenerStart Exception sending context initialized event to listener instance of class [org.springframework.web.context.ContextLoaderListener]
	org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'birtReportService': Invocation of init method failed; nested exception is java.lang.NullPointerException.`

**My Service Class code is** 
 `package service;

import dto.OutputType;
import org.eclipse.birt.core.exception.BirtException;
import org.eclipse.birt.core.framework.Platform;
import org.eclipse.birt.report.engine.api.*;

import org.springframework.beans.factory.DisposableBean;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;
import org.springframework.core.io.ResourceLoader;
import org.springframework.stereotype.Service;
import static dto.OutputType.HTML;
import static dto.OutputType.PDF;
import dto.Report;
import dto.Report.Parameter;
import java.io.File;
import java.util.*;
import javax.annotation.PostConstruct;
import javax.servlet.ServletContext;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

@Service
public class BirtReportService implements ApplicationContextAware, DisposableBean {
	
    @Value("${reports.relative.path}")
    private String reportsPath;
    
    @Value("${images.relative.path}")
    private String imagesPath;

    private HTMLServerImageHandler htmlImageHandler = new HTMLServerImageHandler();

    @Autowired
    private ResourceLoader resourceLoader;
    @Autowired
    private ServletContext servletContext;

    private IReportEngine birtEngine;
    private ApplicationContext context;
    private String imageFolder;

    private Map<String, IReportRunnable> reports = new HashMap<>();

    @SuppressWarnings("unchecked")
    @PostConstruct
    protected void initialize() throws BirtException {
        EngineConfig config = new EngineConfig();
        config.getAppContext().put("spring", this.context);
        Platform.startup(config);
        IReportEngineFactory factory = (IReportEngineFactory) Platform
          .createFactoryObject(IReportEngineFactory.EXTENSION_REPORT_ENGINE_FACTORY);
        birtEngine = factory.createReportEngine(config);
        imageFolder = System.getProperty("user.dir") + File.separatorChar + reportsPath + imagesPath;
        loadReports();
    }

    @Override
    public void setApplicationContext(ApplicationContext context) {
        this.context = context;
    }

    /**
     * Load report files to memory
     *
     */
    public void loadReports() throws EngineException {
        File folder = new File(reportsPath);
        for (String file : Objects.requireNonNull(folder.list())) {
            if (!file.endsWith(".rptdesign")) {
                continue;
            }

            reports.put(file.replace(".rptdesign", ""), birtEngine.openReportDesign(folder.getAbsolutePath() + File.separator + file));

        }
    }

    public List<Report> getReports() {
        List<Report> response = new ArrayList<>();
        for (Map.Entry<String, IReportRunnable> entry : reports.entrySet()) {
            IReportRunnable report = reports.get(entry.getKey());
            IGetParameterDefinitionTask task = birtEngine.createGetParameterDefinitionTask(report);
            Report reportItem = new Report(report.getDesignHandle().getProperty("title").toString(), entry.getKey());
            for (Object h : task.getParameterDefns(false)) {
                IParameterDefn def = (IParameterDefn) h;
                List<Parameter> parameters = reportItem.getParameters();
                if(parameters != null) {
                	reportItem.getParameters()
                		.add(new Report.Parameter(def.getPromptText(), def.getName(), getParameterType(def)));
                }
            }
            response.add(reportItem);
        }
        return response;
    }

    private Report.ParameterType getParameterType(IParameterDefn param) {
        if (IParameterDefn.TYPE_INTEGER == param.getDataType()) {
            return Report.ParameterType.INT;
        }
        return Report.ParameterType.STRING;
    }

    public void generateMainReport(String reportName, OutputType output, HttpServletResponse response, HttpServletRequest request) {
        switch (output) {
        case HTML:
            generateHTMLReport(reports.get(reportName), response, request);
            break;
        case PDF:
            generatePDFReport(reports.get(reportName), response, request);
            break;
        default:
            throw new IllegalArgumentException("Output type not recognized:" + output);
        }
    }

    /**
     * Generate a report as HTML
     */
    @SuppressWarnings("unchecked")
    private void generateHTMLReport(IReportRunnable report, HttpServletResponse response, HttpServletRequest request) {
        IRunAndRenderTask runAndRenderTask = birtEngine.createRunAndRenderTask(report);
        response.setContentType(birtEngine.getMIMEType("html"));
        IRenderOption options = new RenderOption();
        HTMLRenderOption htmlOptions = new HTMLRenderOption(options);
        htmlOptions.setOutputFormat("html");
        htmlOptions.setBaseImageURL("/" + reportsPath + imagesPath);
        htmlOptions.setImageDirectory(imageFolder);
        htmlOptions.setImageHandler(htmlImageHandler);
        runAndRenderTask.setRenderOption(htmlOptions);
        runAndRenderTask.getAppContext().put(EngineConstants.APPCONTEXT_BIRT_VIEWER_HTTPSERVET_REQUEST, request);
        
//        For orders_by_customer report - Set parameter cn
        Map<String, Object> params = new HashMap<>();
        params.put("cn", 112);
        runAndRenderTask.setParameterValues(params);

        try {
            htmlOptions.setOutputStream(response.getOutputStream());
            runAndRenderTask.run();
        } catch (Exception e) {
            throw new RuntimeException(e.getMessage(), e);
        } finally {
            runAndRenderTask.close();
        }
    }

    /**
     * Generate a report as PDF
     */
    @SuppressWarnings("unchecked")
    private void generatePDFReport(IReportRunnable report, HttpServletResponse response, HttpServletRequest request) {
        IRunAndRenderTask runAndRenderTask = birtEngine.createRunAndRenderTask(report);
        response.setContentType(birtEngine.getMIMEType("pdf"));
        IRenderOption options = new RenderOption();
        PDFRenderOption pdfRenderOption = new PDFRenderOption(options);
        pdfRenderOption.setOutputFormat("pdf");
        runAndRenderTask.setRenderOption(pdfRenderOption);
        runAndRenderTask.getAppContext().put(EngineConstants.APPCONTEXT_PDF_RENDER_CONTEXT, request);
        
//      For orders_by_customer report - Set parameter cn
      Map<String, Object> params = new HashMap<>();
      params.put("cn", 112);
      runAndRenderTask.setParameterValues(params);

        try {
            pdfRenderOption.setOutputStream(response.getOutputStream());
            runAndRenderTask.run();
        } catch (Exception e) {
            throw new RuntimeException(e.getMessage(), e);
        } finally {
            runAndRenderTask.close();
        }
    }
    
    @Override
    public void destroy() {
        birtEngine.destroy();
        Platform.shutdown();
    }
}
`
 i am using Netbean 8.2 with jdk 1.8 and Runtime for Birt jar file is org.eclipse.birt.runtime_4.8.0-20180626-4.8.0 (1).jar

**Comment by speckyspooky** [comment]
What means the debugger? At which line appears the exception with the NPE?

**Reply by irfankundi786** [reply]
log shows that error is occur at @Service class autowire exception here is log
`08-Jun-2024 21:06:11.763 INFO [http-nio-8080-exec-531] org.apache.catalina.core.ApplicationContext.log Destroying Spring FrameworkServlet 'dispatcher' 08-Jun-2024 21:06:11.764 INFO [http-nio-8080-exec-531] org.apache.catalina.core.ApplicationContext.log Closing Spring root WebApplicationContext 08-Jun-2024 21:06:27.899 INFO [http-nio-8080-exec-530] org.apache.catalina.core.ApplicationContext.log No Spring WebApplicationInitializer types detected on classpath 08-Jun-2024 21:06:27.931 INFO [http-nio-8080-exec-530] org.apache.catalina.core.ApplicationContext.log Initializing Spring root WebApplicationContext 08-Jun-2024 21:06:32.049 SEVERE [http-nio-8080-exec-530] org.apache.catalina.core.StandardContext.listenerStart Exception sending context initialized event to listener instance of class [org.springframework.web.context.ContextLoaderListener] org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'birtReportService': Invocation of init method failed; nested exception is java.lang.NullPointerException.`

**Reply by merks** [reply]
Without a stack trace, no one can know where exactly the null point exception happens so there is really no information here that can be used to diagnose the details of this problem.

**Reply by irfankundi786** [reply]
which detail you needed i can past here ? console error messages ?

**Reply by merks** [reply]
Here is an example of an issue that was opened with a stack trace:

https://github.com/eclipse-birt/birt/issues/1711

I would expect and hope that whatever logging system is logging information includes not just a textual messages about an exception but also the logs the actual stack trace of the exception.  From that we can know which line of which class has a problem...

**Comment by irfankundi786** [comment]
09-Jun-2024 15:44:52.056 INFO [http-nio-8080-exec-167] org.apache.catalina.core.ApplicationContext.log No Spring WebApplicationInitializer types detected on classpath
09-Jun-2024 15:44:52.086 INFO [http-nio-8080-exec-167] org.apache.catalina.core.ApplicationContext.log Initializing Spring root WebApplicationContext
09-Jun-2024 15:44:55.298 SEVERE [http-nio-8080-exec-167] org.apache.catalina.core.StandardContext.listenerStart Exception sending context initialized event to listener instance of class [org.springframework.web.context.ContextLoaderListener]
	org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'birtReportService': Invocation of init method failed; nested exception is java.lang.NoClassDefFoundError: org/eclipse/core/runtime/CoreException
		at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:136)
		at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:407)
		at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1545)
		at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:539)
		at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:475)
		at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:304)
		at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:228)
		at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:300)
		at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:195)
		at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:700)
		at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:760)

**Reply by irfankundi786** [reply]
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'birtReportService': Invocation of init method failed; nested exception is java.lang.NoClassDefFoundError: org/eclipse/core/runtime/CoreException

**Reply by merks** [reply]
That suggests you have jars missing from the classpath.  The CoreException class is in some version of this jar:

![image](https://github.com/eclipse-birt/birt/assets/208716/cc846a91-9155-402d-9cfd-8a6001b9e579)

**Reply by irfankundi786** [reply]
But unable to configure which jar file is missing ?  I am using NetBeans with jdk 8.2 and want to run Birt Report from Spring Project. Already include org.eclipse.birt.runtime_4.8.0-20180626-4.8.0 (1).jar     jar file for Birt Report

**Reply by merks** [reply]
The runtime zip contains a whole whack of jars. 

![image](https://github.com/eclipse-birt/birt/assets/208716/65ed5646-9a87-4822-a01f-e6fddb49205e)

I don't think the one jar is enough.  I'm not sure what the minimal set of jars is...

**Reply by irfankundi786** [reply]
i have already installed org.eclipse.birt.runtime_4.8.0-20180626-4.8.0 (1).jar

**Reply by irfankundi786** [reply]
but not known whether it is compatible with JDK1.8 or not ?

**Reply by merks** [reply]
You mentioned using version 4.8 so presumably you have this:

![image](https://github.com/eclipse-birt/birt/assets/208716/2666b63a-1008-4c92-999d-688213f06696)

and all those things are 1.8 compatible for that old version.

**Reply by irfankundi786** [reply]
Not working for me with this jar file @merks

**Reply by irfankundi786** [reply]
I am trying with different version not sure about which version is compatible with JDK1.8

**Reply by irfankundi786** [reply]
if you want i can share the project source code

**Reply by merks** [reply]
I don't know exactly what you downloaded.  Presumably some older version from here if you need to run on Java 1.8:

https://archive.eclipse.org/birt/downloads/drops/

After that, I don't know exactly what you are trying.  But if you look in the war file for example, you see also a whole whack of jars:

![image](https://github.com/eclipse-birt/birt/assets/208716/dc79a3b8-88cb-4e6a-a0e2-8c4c3b6474bb)

So pick something compatible with Java 1.8 and the necessary jars; as I said, I'm not sure that a subset will do the trick and if so which subset.

**Reply by merks** [reply]
It sounds more like a problem with how you have configured the jars to be included in the launched application. I would have no clue how to launch a spring application, nor would I have the time to spend on helping with such a thing. If you figured out how to add the one birt jar, surely adding all the other jars can't be such a big challenge?

**Comment by irfankundi786** [comment]
I am using spring MVC to call birt report.But stuck to resolve relative path like

`@Value("${reports.relative.path}")
    private String reportsPath;
  @Value("${images.relative.path}")
    private String imagesPath;`

I have created file with name of application.properties under the path of /WEB-INF/RESOURSES/APPLICATION.PROPERTIES and this file contains entries like
`reports.relative.path=your/reports/path
images.relative.path=your/images/path`


my application.context.xml entry as
`
<mvc:resources mapping="/reports/**" location="/WEB-INF/reports/" />

<mvc:resources mapping="/images/**" location="/WEB-INF/images/" />`


but when i want to load this in my service class this not resolved

Actually i want to call the birt report from spring mvc and for this purpose i have make all the configuration changes but there is something i am missing thats why not able to resolve the path of report and images as i configured and tell above. for clarity i am pasted here my service class code

**Reply by speckyspooky** [reply]
I'm cannot assist with the spring-framwork.
But I didn't understand your configuration of BIRT because the default path of the report-folder is another one at the default BIRT-webapp package and alsi the WEB-INF-folder contains other folders.

The screens based on BIRT-OSGi-version. Do you have a completely different structure...?

My report-folder, default of BIRT:
- birt/report
- birt/report/images
![grafik](https://github.com/eclipse-birt/birt/assets/41593722/a71eb571-22f5-4fec-aa6f-cc979e96e2f1)

My WEB-INF-folder, default of BIRT:
![grafik](https://github.com/eclipse-birt/birt/assets/41593722/dedc1a94-61e2-4c37-9592-1b53dd689a13)

**Comment by SteveSchafer-Innovent** [comment]
A few years ago we created a sample spring boot BIRT application which you can find here: https://github.com/innoventsolutions/birt-spring-boot.
This uses the BIRT runtime library, not the webapp.  Maybe you can find something useful here.
